### ISSUE TITLE
Grid search with Horovod and Keras

### ISSUE THREAD

**Issue by pbruneau** [horovod/horovod#1766]
**Environment:**
1. Framework: Keras
2. Framework version: 2.2.4
3. Horovod version: 0.19.0
4. MPI version: 3.0.0
5. CUDA version: 10.0
6. NCCL version: NA
7. Python version: 3.6.3
8. OS and version: CentOS Linux 7
9. GCC version: 7.2.0

**Your question:**

I would like to combine grid search with using horovod, basically nesting a piece of code such as https://github.com/horovod/horovod/blob/master/examples/keras_mnist.py (denoted as `training(p)` below), for example:

    for p in params:
        model = training(p)
        # then compute and store some performance metric for model

However, as far as I understood, `horovodrun` boils down to spawning one process per GPU via mpirun, and sync between processes is performed through `hvd.init()` and at the optimizer level.

What would be the good practice to make horovod able to perform this synchronization repeatedly within my for loop? Would launching my script with `horovodrun` work out of the box?

**Comment by tgaddair** [comment]
Hey @pbruneau, I would suggest using [horovod.run.run](https://horovod.readthedocs.io/en/latest/api.html#module-horovod.run) for this.  This will allow you to spawn a separate Horovod distributed job for every call to `training(p)`.  

Some care needs to be taken to handle proper serialization/deserialization of the model over pickle.  For that, I would take a look at [this](https://github.com/horovod/horovod/blob/master/examples/keras_spark_rossmann_run.py#L342) example.  

Let me know if you have any questions.

**Comment by pbruneau** [comment]
Hi @tgaddair, this seems to suit my needs indeed, thanks!

So if I understand correctly, I would call the enclosing script the regular way (i.e. `python3 script.py`), call `training(p)` with `horovod.run.run`, and have the individual `hvd.init()` etc... in the training function, right?

Also, I use horovod in the context of a slurm batch script on a HPC, with MPI and Tensorflow 1.14.0 (cf Troubleshooting section), so currently I use the following command for a "single model" script:

    LD_PRELOAD=/usr/lib64/libhwloc.so mpirun -np 2 \
        -npernode 2 -bind-to none \
        -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH \
        -mca pml ob1 -mca btl ^openib \
        python3 script.py --horovod_device=gpu

(for a job on 2 nodes and 2 gpus per node). Does this mean that I would to cast this command to the `use_mpi` and `mpi_args` parameters of the `horovod.run.run` function + set `LD_PRELOAD` using `os.environ` (hoping that this would actually work)?

**Comment by pbruneau** [comment]
To update the above: I tried to run a SLURM script that books 2 nodes (with 2 GPU per node) and launches `script.py`. So quite logically, a single process is spawned in one of the nodes (say `gpu1`). I would like to use all available 4 GPU.

In `script.py`, I tried to hack my way around (basically customize the `mpirun_command` in `mpi_run.py` to match the code block above), but the best I could get has been to use the 2 GPU of `gpu1`. In brief, I did not find the way to tell `mpirun` the way to find both nodes assigned to the task - for it, only `gpu1` seems to be there.
